spyder_kernels/comms/commbase.py

```diff
@@ -0,0 +1,464 @@
+        self._message_handlers = {}
+        self._remote_call_handlers = {}
+        # Lists of reply numbers
+        self._reply_inbox = {}
+        self._reply_waitlist = {}
+
+        self._register_message_handler(
+            'remote_call', self._handle_remote_call)
+        self._register_message_handler(
+            'remote_call_reply', self._handle_remote_call_reply)
+
+        # Dummy functions for testing and to trigger side effects such as
+        # an interruption or waiting for a reply.
+        def pong_back():
+            self.remote_call(self.calling_comm_id).pong()
+
+        self.register_call_handler('ping', pong_back)
+        self.register_call_handler('pong', lambda: None)
+
+    def close(self, comm_id=None):
+        """Close the comm and notify the other side."""
+        if comm_id is None:
+            # close all the comms
+            id_list = list(self._comms.keys())
+        else:
+            id_list = [comm_id]
+
+        for comm_id in id_list:
+            self._comms[comm_id].close()
+            del self._comms[comm_id]
+
+    def is_open(self, comm_id=None):
+        """Check to see if the comm is open."""
+        if comm_id is None:
+            return len(self._comms) > 0
+        return comm_id in self._comms
+
+    def register_call_handler(self, call_name, handler):
+        """
+        Register a remote call handler.
+
+        Parameters
+        ----------
+        call_name : str
+            The name of the called function.
+        handler : callback
+            A function to handle the request, or `None` to unregister
+            `call_name`.
+        """
+        if not handler:
+            self._remote_call_handlers.pop(call_name, None)
+            return
+
+        self._remote_call_handlers[call_name] = handler
+
+    def remote_call(self, comm_id=None, callback=None, **settings):
+        """Get a handler for remote calls."""
+        return RemoteCallFactory(self, comm_id, callback, **settings)
+
+    # ---- Private -----
+    def _send_message(self, spyder_msg_type, content=None, data=None,
+                      comm_id=None):
+        """
+        Publish custom messages to the other side.
+
+        Parameters
+        ----------
+        spyder_msg_type: str
+            The spyder message type
+        content: dict
+            The (JSONable) content of the message
+        data: any
+            Any object that is serializable by cloudpickle (should be most
+            things). Will arrive as cloudpickled bytes in `.buffers[0]`.
+        comm_id: int
+            the comm to send to. If None sends to all comms.
+        """
+        if not self.is_open(comm_id):
+            raise CommError("The comm is not connected.")
+        import cloudpickle
+        msg_dict = {
+            'spyder_msg_type': spyder_msg_type,
+            'content': content,
+            'PY2': PY2,
+            }
+        buffers = [cloudpickle.dumps(data, protocol=PICKLE_PROTOCOL)]
```

Comment:
I think we're not handling correctly the case of a failure in serializing the requested data here. Instead of generating a traceback due to that (which is shown in our error dialog now), we need to catch `PicklingError` (or perhaps any exception?) and make `buffers` content return something that can be interpreted correctly on the other side.

This is quite similar to the way you're handling failures in `cloudpickle.loads` below.

Reply:
I disagree: There are two kinds of errors:
 - Errors caused by the same side, which can be raised as is.
 - Errors caused by the other side, that have to be caught and send back.

Here, either the first point applies, or the exception is caught in `_handle_remote_call`.

Reply:
So for example if the failure happens because the frontend requested data that could not be pickled, the error generated here will be sent to the frontend. Is it not what you see?

Reply:
By running this code, I just have one QWindows showing up.
```
try:
    1/0
except:
    import sys
    tb = sys.exc_info()[2]
var = [tb]
```

Reply:
![Screenshot 2019-08-09 at 16 47 13](https://user-images.githubusercontent.com/6740194/62791670-84bfd980-bac5-11e9-9fc7-fecb0368b06d.png)

Reply:
> Is it not what you see?

Yes, that's what I'm seeing. But I don't want users to report it in case they are trying to visualize a non-picklable object in the Variable Explorer.

A very simple example:

```python
In [1]: import sys

In [2]: f = sys._getframe()
```

If you double-click in `f` in the Variable Explorer, you'll get:

![Selección_016](https://user-images.githubusercontent.com/365293/62791668-e4ba7e00-bacd-11e9-939b-ac51819d8ef9.png)

Showing the QMessageBox in the front is what we were doing before, and that's fine. But showing the error dialog is completely unnecessary because frames are not picklable.

Reply:
I agree but I am not sure why you see the error dialog. It doesn't appear on my laptop.

Reply:
![Screenshot 2019-08-09 at 17 05 52](https://user-images.githubusercontent.com/6740194/62792824-fdc03080-bac7-11e9-867b-2364595a7977.png)

Reply:
Also f is not in the variable explorer?

Reply:
I would be templted to say that this is out of the scope of this PR.

Reply:
What is the error and traceback in the  error dialog you have?

Reply:
Or even simpler: you could open a new issue with the error dialog so I can see the traceback and other infos :p

Reply:
This was probably fixed with your switch to use Pickle HIGHEST_PROTOCOL, because I'm not seeing the error dialog shown anymore (only the QMessageBox).

So thanks for that!